### ISSUE TITLE
Accesing WebAssembly Standalone function from Javascript

### ISSUE THREAD

**Issue by zzarcon** [kripken/emscripten#4844]
I'm trying to call an exported C function from javascript, but I can't figure out how to force emscripten to include the needed libraries:

**emcc command**
```
$ emcc main.c -Os -s WASM=1 -s SIDE_MODULE=1 -o main.wasm
```

**main.c**
```c
#include <string.h>

size_t getLength(char* str) {
  size_t len = strlen(str);
  
  return len;
}
```
**app.js**
```javascript
function loadModule(filename) {
  return fetch(filename)
    .then(response => response.arrayBuffer())
    .then(buffer => WebAssembly.compile(buffer))
    .then(module => {
      const imports = {
        env: {
          memoryBase: 0,
          tableBase: 0,
          memory: new WebAssembly.Memory({
            initial: 256
          }),
          table: new WebAssembly.Table({
            initial: 0,
            element: 'anyfunc'
          })
        }
      };
      
      return new WebAssembly.Instance(module, imports);
    });
}

loadModule('main.wasm').then(instance => {
      const getLength = instance.exports._getLength;

      console.log(getLength("foo"));
});
```

I get the following error after running the app.js:

```
Uncaught (in promise) LinkError: WebAssembly.Instance(): Import #1 module="env" function="_strlen" error: function import requires a callable
```

I figure out that this happen because **main.wast** is trying to call **_strlen** from **env**:

```
(import "env" "_strlen" (func $_strlen (param i32) (result i32)))
```

So I thought about passing the function to the env to the previous **imports**, just like:

```javascript
const imports = {env: {_strlen: function(str) {}}}
```

That makes the code to call the passed **_strlen** function and doesn't break, but is not sending any value inside **str**.

I also tried to pass the EMCC_FORCE_STDLIBS when building:

```
$ EMCC_FORCE_STDLIBS=1 emcc main.c -Os -s WASM=1 -s SIDE_MODULE=1 -o main.wasm
```
But the result is the same 😅 .

Pretty much all the code in this issue comes from @kripken [gist](https://gist.github.com/kripken/59c67556dc03bb6d57052fedef1e61ab), so maybe you can tell what Im doing wrong :)

Thanks!

**Comment by kripken** [comment]
This is how side modules work: they [don't have system libs like libc linked in](https://github.com/kripken/emscripten/wiki/Linking#overview-of-dynamic-linking). The reason is that system libs can be comprised of JS as well, which would mean the wasm is no longer standalone.

If all you need is strlen, you can implement it in JS as you wrote - I don't know why that failed? What specifically went wrong? (Note that the input and output are i32s, not JS strings.)

Another option for strlen is to just include it and have it compiled manually. For example, include the strlen.c source file with your own source file that you compile.

**Comment by zzarcon** [comment]
What failed when I pass the **_strlen** in the environment is that the **str** was 0:

![screen shot 2017-01-08 at 23 03 57](https://cloud.githubusercontent.com/assets/1194982/21753931/c9729498-d5f6-11e6-8c09-7d48934e61f1.png)

I was calling it with:
```javascript
loadModule('main.wasm').then(instance => {
      const getLength = instance.exports._getLength;
      console.log(getLength("foo"));
});
```
Is that maybe because I have something wrong configured in the env?

I'd like to go with the first approach you mention, like passing the JS implementation and manipule the i32s, can you show an example of how this will work? including the i32s manipulation if possible. Thanks

**Comment by kripken** [comment]
Everything you are doing is ok, except for the fact that you're passing in a JS string when you do `getLength("foo")`. The wasm method expects an i32 (representing a C pointer) so it's turned into an i32, namely 0 (same as if you do `"foo" | 0` in JS, it casts the string to a JS Number, 0). Then the compiled code calls `strlen` on 0, which is why you see that 0 there.

What you want to do is pass in a C pointer instead of a JS string. That's not trivial to do with standalone wasm. Emscripten's normal JS glue, for non-standalone code, can do this for you, e.g. if you use ccall/cwrap. But you can also do it manually in standalone code - you need to allocate room for the string in C memory, write it there, then pass that pointer to `getLength`, then free the memory.

**Comment by zzarcon** [comment]
Alright I got it now, I was missing the whole emscripten [ccall/cwrap](https://kripken.github.io/emscripten-site/docs/api_reference/preamble.js.html#cwrap) part, sorry for that, I just came here from the WebAssembly docs and I was not sure how all the pieces are connected.

This works fine as you said:

```javascript
const getLength = Module.cwrap('getLength', 'number', ['string']); 

getLength("foo") == 3
```

Also It was not clear to me in the beginning where `Module` was coming from, until I realized its generated when you compile the program without `SIDE_MODULE=1`.

Last question, is this `Module` something generic that I can reuse with any other program or it contains info based on the program that was compiled? Just asking from a user point of view cause I think will be great if we could provide a client-side api through a npm package like:

```javascript
import { cwrap } from 'emscripten';

const getLength = cwrap('anyMethodName', 'number', ['string']); 
//...
```

Thanks for the great job and explanation before!
